solvers/H3LAPD/EquationSystems/HW2Din3DSystem.cpp

```diff
@@ -0,0 +1,190 @@
+#include <LibUtilities/BasicUtils/Vmath.hpp>
+#include <LibUtilities/TimeIntegration/TimeIntegrationScheme.h>
+#include <boost/core/ignore_unused.hpp>
+
+#include "HW2Din3DSystem.hpp"
+
+namespace NESO::Solvers::H3LAPD {
+std::string HW2Din3DSystem::class_name =
+    SU::GetEquationSystemFactory().RegisterCreatorFunction(
+        "2Din3DHW", HW2Din3DSystem::create,
+        "(2D) Hasegawa-Wakatani equation system as an intermediate step "
+        "towards the full H3-LAPD problem");
+
+HW2Din3DSystem::HW2Din3DSystem(const LU::SessionReaderSharedPtr &session,
+                               const SD::MeshGraphSharedPtr &graph)
+    : UnsteadySystem(session, graph), AdvectionSystem(session, graph),
+      DriftReducedSystem(session, graph) {
+  m_required_flds = {"ne", "w", "phi"};
+  m_int_fld_names = {"ne", "w"};
+
+  // Frequency of growth rate recording. Set zero to disable.
+  m_diag_growth_rates_recording_enabled =
+      session->DefinesParameter("growth_rates_recording_step");
+
+  // Frequency of mass recording. Set zero to disable.
+  m_diag_mass_recording_enabled =
+      session->DefinesParameter("mass_recording_step");
+}
+
+/**
+ * @brief Override DriftReducedSystem::calc_E_and_adv_vels in order to set
+ * electron advection veloctity in v_ExB
+ *
+ * @param in_arr array of field phys vals
+ */
+void HW2Din3DSystem::calc_E_and_adv_vels(
+    const Array<OneD, const Array<OneD, NekDouble>> &in_arr) {
+  DriftReducedSystem::calc_E_and_adv_vels(in_arr);
+  int npts = GetNpoints();
+
+  Vmath::Zero(npts, m_par_vel_elec, 1);
+  // vAdv[iDim] = b[iDim]*v_par + v_ExB[iDim] for each species
+  for (auto iDim = 0; iDim < m_graph->GetSpaceDimension(); iDim++) {
+    Vmath::Svtvp(npts, m_b_unit[iDim], m_par_vel_elec, 1, m_ExB_vel[iDim], 1,
+                 m_adv_vel_elec[iDim], 1);
+  }
+}
+
+/**
+ * @brief Populate rhs array ( @p out_arr ) for explicit time integration of
+ * the 2D Hasegawa Wakatani equations.
+ *
+ * @param in_arr physical values of all fields
+ * @param[out] out_arr output array (RHSs of time integration equations)
+ */
+void HW2Din3DSystem::explicit_time_int(
+    const Array<OneD, const Array<OneD, NekDouble>> &in_arr,
+    Array<OneD, Array<OneD, NekDouble>> &out_arr, const NekDouble time) {
+
+  // Check in_arr for NaNs
+  for (auto &var : {"ne", "w"}) {
+    auto fidx = m_field_to_index.get_idx(var);
+    for (auto ii = 0; ii < in_arr[fidx].size(); ii++) {
+      if (!std::isfinite(in_arr[fidx][ii])) {
+        std::cout << "NaN in field " << var << ", aborting." << std::endl;
+        exit(1);
```

Comment:
Is there any intercepting of this in Nektar++ I'm not aware of to handle this in parallel? There is a helper function in NESO-Particles `NESOASSERT(condition, message)` to throw error messages. What should happen (and I notice I've missed this in NESOASSERT - should update) is that the message is thrown then MPI_Abort is called. If you simply kill one MPI rank then the remaining ranks will block on a barrier until the job walltime runs out.

Reply:
Nope. Agree this definitely needs to be changed.
Happy to use a NESOASSERT and have it call MPI_Abort, although it would be good to have a version that was independent of NP...